interop-testing/src/test/java/io/grpc/testing/integration/NettyFlowControlTest.java

```diff
@@ -0,0 +1,270 @@
+  }
+
+  @Test
+  public void highBandHighLatency() throws InterruptedException, ExecutionException {
+    resetProxy(HIGH_BAND, HIGH_LAT, TimeUnit.MILLISECONDS).get();
+    resetConnection(REGULAR_WINDOW, REGULAR_WINDOW);
+    doTest(HIGH_BAND, HIGH_LAT);
+  }
+
+  @Test
+  public void verySmallWindow() throws InterruptedException, ExecutionException {
+    resetProxy(MED_BAND, MED_LAT, TimeUnit.MILLISECONDS).get();
+    resetConnection(REGULAR_WINDOW, TINY_WINDOW);
+    doTest(MED_BAND, MED_LAT);
+  }
+
+  /**
+   * Main testing method. Streams 2 MB of data from a server and records the final window and
+   * average bandwidth usage.
+   */
+  private void doTest(int bandwidth, int latency) throws InterruptedException {
+
+    int streamSize = 2 * 1024 * 1024;
+
+
+    TestServiceGrpc.TestServiceStub stub = TestServiceGrpc.newStub(channel);
+    StreamingOutputCallRequest.Builder builder = StreamingOutputCallRequest.newBuilder();
+    builder.addResponseParameters(ResponseParameters.newBuilder().setSize(streamSize));
+    StreamingOutputCallRequest request = builder.build();
+
+    TestStreamObserver observer = new TestStreamObserver();
+    stub.streamingOutputCall(request, observer);
+    observer.waitFor();
+
+    long expectedWindow = (latency * (bandwidth / TimeUnit.SECONDS.toMillis(1))) * 2;
+    int lastWindow = HandlerSettings.getLatestClientWindow();
+
+    // deal with cases that either don't cause a window update or hit max window
+    expectedWindow = Math.min(MAX_WINDOW, (Math.max(expectedWindow, REGULAR_WINDOW)));
+
+    // Range looks large, but this allows for only one extra/missed window update
+    // (one extra update causes a 2x difference and one missed update causes a .5x difference)
+    assertTrue(lastWindow < 2 * expectedWindow);
+    assertTrue(expectedWindow < 2 * lastWindow);
+  }
+
+  /**
+   * Resets client/server and their flow control windows.
+   */
+  private void resetConnection(int serverFlowControlWindow, int clientFlowControlWindow)
+      throws InterruptedException {
+    if (channel != null) {
+      if (!channel.isShutdown()) {
+        channel.shutdown();
+        channel.awaitTermination(100, TimeUnit.MILLISECONDS);
+      }
+    }
+    startServer(serverFlowControlWindow);
+    int channelPort = useProxy ? proxyPort : port;
+    channel = NettyChannelBuilder.forAddress(new InetSocketAddress("localhost", channelPort))
+        .flowControlWindow(clientFlowControlWindow)
+        .negotiationType(NegotiationType.PLAINTEXT)
+        .build();
+  }
+
+  private void startServer(int serverFlowControlWindow) {
+    ServerBuilder<?> builder =
+        NettyServerBuilder.forAddress(new InetSocketAddress("localhost", port))
+        .flowControlWindow(serverFlowControlWindow);
+    List<ServerInterceptor> allInterceptors = ImmutableList.of();
+    builder.addService(ServerInterceptors.intercept(
+        new TestServiceImpl(Executors.newScheduledThreadPool(2)),
+        allInterceptors));
```

Comment:
That works. I just now saw this comment after I committed so I'll include the change when I squash and push, but I want to make sure the builds pass first.